### PR TITLE
Added various Harvest Temple mechanics and decorations, new APIs

### PR DESCRIPTION
**Pull request by Linkaaaaa**
# Harvest Temple

## Mechanics
- Hit by Void Explosion (Last Laugh)
- Hit by Magic Discharge
- Hit by Lightning of Jormag, Flames of Primordus, Stormfall (Purification 1)
- Hit by Grasp of Jormag
- Hit by Timecaster's Gravity Crush and Nightmare Epoch
- Changed Skullpiercer's Kick to check for Stability
- Hit by Zhaitan's Tail Slam
- Hit by Corrupted Waters (Purification 3)
- Hit by Saltspray's Hydro Burst, Call Lightning, Frozen Fury, Rolling Flame, Shatter Earth
- Hit by Grasp of the Void (Purification 4)
- Hit by Obliterator's Firebomb, Wyvern Breath, Charge (hit and cc)
- Hit by Goliath's Glacial Slam (hit and cc)

## Decorations
- Changed purification zone from blue to white
- Changed Lightning of Jormag from yellow to blue
- Added Brandbombers' projectile animation and indicator
- Added Kralkatorrik's Crystal Barrage
- Added Timecaster's Gravity Ball (warning indicator and moving gadget) and Nightmare Epoch
- Added Mordremoth's Shockwaves
- Added Zhaitan's Tail Slam and Zhaitan's Scream (NM and CM)
- Added Giants' Death Scream and Rotting Bile
- Adjusted Zhaitan's Reach pull doughnut radius
- Added Saltspray's Hydro Burst, Call Lightning, Frozen Fury, Rolling Flame, Shatter Earth
- Adjusted various conflicting effects for Soo-Won and other mechanics
- Adjusted Torment of the Void orb AoEs duration for new and old logs
- Added projectile animations for Tsunami Slam, Tail Slam, Claw Slap
- Added Obliterator's Firebomb, Wyvern Breath, Charge, Shockwave
- Added Goliath's Glacial Slam
- Added Abomination's Swipe

## General
- Redirected Grasp of Jormag and Frost Meteor to Jormag
- Redirected Zhaitan's Tail Slam and Scream of Zhaitan CM to Zhaitan
- Redirected Tail Slam to Soo-Won
- Added Gravity Ball NPC
- Changed icon for the killable orb
- Added various comments for documentation and cleanup

# General

## API
- Created AddProjectile in CombatReplay
- Created ComputeEndCastTimeByStun EncounterLogicUtils (will port over the fractal encounters to this API in another PR)

## Misc
- Updated Light Orange color to the old AoE color we used (250, 120, 0)
- Added Ice color

### CODE REVIEW

File: GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs [hunk 1]
```diff
@@ -452,6 +478,22 @@ internal override void EIEvtcParse(ulong gw2Build, int evtcVersion, FightData fi
             {
                 agentData.Refresh();
             }
+            // Gravity Ball - Timecaster gadget
+            var gravityBalls = combatData.Where(x => x.DstAgent == 0 && x.IsStateChange == ArcDPSEnums.StateChange.MaxHealthUpdate).Select(x => agentData.GetAgent(x.SrcAgent, x.Time)).Where(x => x.Type == AgentItem.AgentType.Gadget && x.HitboxHeight == 300 && x.HitboxWidth == 100 && x.Master == null).ToList();
```
Comment (EliphasNUIT) [inline GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs:482]: This block should be before agentData.Refresh()

Comment (Linkaaaaa) [reply GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs:482]: I was moving things around and didn't notice i moved it under the refresh :x

File: GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs [hunk 2]
```diff
@@ -1157,5 +1799,45 @@ private static bool CustomCheckVoidwalkerEligibility(ParsedEvtcLog log)
...
+        /// <param name="endEffects">List of the final effects for the positions.</param>
+        /// <returns>Filtered list with matched <paramref name="startEffects"/>, <paramref name="endEffects"/> and distance between them.</returns>
+        private static List<(EffectEvent, EffectEvent, float)> MatchEffectToEffect(IReadOnlyList<EffectEvent> startEffects, IReadOnlyList<EffectEvent> endEffects)
+        {
+            var matchedEffects = new List<(EffectEvent, EffectEvent, float)>();
+            foreach (EffectEvent startEffect in startEffects)
+            {
+                float minimalDistance = 0;
+                EffectEvent matchedEffect = null;
+                foreach (EffectEvent endEffect in endEffects.Where(x => x.Time > startEffect.Time + 200 && Math.Abs(x.Time - startEffect.Time) < 10000))
```
Comment (EliphasNUIT) [inline GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs:1816]: candidateEffectEvents = endEffects.Where(x => x.Time > startEffect.Time + 200 && Math.Abs(x.Time - startEffect.Time) < 10000).ToList();
if (candidateEffectsEvents.Any()) {
var matchedEffect = candidateEffectEvents.MinBy(x => x.Position.Distance2DToPoint(startEffect.Position);
minimalDistance = matchedEffect.Position.DIstance2DToPoint(startEffect.Position);
matchedEffects.Add((matchedEffect, startEffect, minimalDistance));
}

File: GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs [hunk 3]
```diff
@@ -1157,5 +1799,45 @@ private static bool CustomCheckVoidwalkerEligibility(ParsedEvtcLog log)
...
+
+        /// <summary>
+        /// Matches an effect to another effect by proximity and filters out additional effects.
+        /// </summary>
+        /// <param name="startEffects">List of the initial effects for the positions.</param>
+        /// <param name="endEffects">List of the final effects for the positions.</param>
+        /// <returns>Filtered list with matched <paramref name="startEffects"/>, <paramref name="endEffects"/> and distance between them.</returns>
+        private static List<(EffectEvent, EffectEvent, float)> MatchEffectToEffect(IReadOnlyList<EffectEvent> startEffects, IReadOnlyList<EffectEvent> endEffects)
+        {
+            var matchedEffects = new List<(EffectEvent, EffectEvent, float)>();
```
Comment (EliphasNUIT) [inline GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs:1811]: added names to the tuple: startEffect, endEffect, distance
Would be much clearer to read than Item1, Item2 and Item3

File: GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs [hunk 4]
```diff
@@ -892,44 +992,203 @@ internal override void ComputeNPCCombatReplayActors(NPC target, ParsedEvtcLog lo
...
+                                // Add projectile
+                                replay.AddProjectile(tailSlamEffect.Position, orbAoeEffect.Position, lifespanAoE, Colors.Black, 0.5);
+                            }
+                        }
+                    }
+
+                    // Tsunami Slam AoE indicator
+                    if (log.CombatData.TryGetEffectEventsByGUID(EffectGUIDs.HarvestTempleSooWonTsunamiSlamIndicator, out IReadOnlyList<EffectEvent> tsunamiSlamIndicators))
+                    {
+                        foreach (EffectEvent clawSlamEffect in tsunamiSlamIndicators)
```
Comment (EliphasNUIT) [inline GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs:1168]: clawSlam? Wrong naming?

File: GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs [hunk 5]
```diff
@@ -892,44 +992,203 @@ internal override void ComputeNPCCombatReplayActors(NPC target, ParsedEvtcLog lo
...
+                                // AoE to AoE effects
+                                foreach ((EffectEvent, EffectEvent, float) match in aoeToAoeMatches)
+                                {
+                                    EffectEvent endingAoE = match.Item1;
+                                    EffectEvent startingAoE = match.Item2;
+                                    long endingAoEDuration = 0;
+                                    long startingAoEDuration = 0;
+                                    foreach (ParametricPoint3D point in positions)
+                                    {
+                                        if (endingAoEDuration != 0 && startingAoEDuration != 0) { break; }
```
Comment (EliphasNUIT) [inline GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs:1072]: Line break, ifs shouldn't be in a single line.

File: GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs [hunk 6]
```diff
@@ -963,6 +1222,378 @@ internal override void ComputeNPCCombatReplayActors(NPC target, ParsedEvtcLog lo
...
+                            var positionConnector = new PositionConnector(effect.Position);
+                            var circleIndicator = new CircleDecoration(radius, (lifespan.start - 2000, lifespan.start), Colors.LightOrange, 0.2, positionConnector);
+                            var circleWhirlpool = new CircleDecoration(radius, lifespan, Colors.LightBlue, 0.2, positionConnector);
+                            replay.AddDecorationWithGrowing(circleIndicator, lifespan.start);
+                            replay.Decorations.Add(circleWhirlpool);
+                            // The whirlpools increase in size every set of 3, find if there is a next effect within 500ms.
+                            EffectEvent nextWhirlpool = hydroBurstWhirlpools.FirstOrDefault(x => Math.Abs(x.Time - effect.Time) < 500 && x.Time > effect.Time);
+                            radius = counter % 3 == 0 ? radius + 10 : radius;
+                            // if there isn't a next one, reset the radius to the starting value
+                            if (nextWhirlpool == null) { radius = 90; }
```
Comment (EliphasNUIT) [inline GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs:1365]: Line break, ifs shouldn't be single line.

File: GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs [hunk 7]
```diff
@@ -963,6 +1222,378 @@ internal override void ComputeNPCCombatReplayActors(NPC target, ParsedEvtcLog lo
...
+                                case 7:
+                                    innerRadius = 100;
+                                    outerRadius = 150;
+                                    break;
+                                default:
+                                    break;
+                            }
+                            var doughnut = new DoughnutDecoration(innerRadius, outerRadius, lifespan, Colors.LightOrange, 0.2, positionConnector);
+                            replay.AddDecorationWithGrowing(doughnut, lifespan.end);
+                            if (counter < 7) { counter++; } else { counter = 1; }
```
Comment (EliphasNUIT) [inline GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs:1447]: Line break, ifs shouldn't be single line.

File: GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs [hunk 8]
```diff
@@ -862,6 +929,17 @@ internal override void ComputeNPCCombatReplayActors(NPC target, ParsedEvtcLog lo
                             replay.AddDecorationWithGrowing(new CircleDecoration( 200, (start, end), "rgba(49, 71, 0, 0.2)", new PositionConnector(effect.Position)), end);
                         }
                     }
+                    // Shockwaves
+                    if (log.CombatData.TryGetEffectEventsByGUID(EffectGUIDs.HarvestTempleMordremothShockwave1, out IReadOnlyList<EffectEvent> shockwaves))
```
Comment (EliphasNUIT) [inline GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs:933]: No Src check? There are a log of Void Amalgamate agents in the fight, every each one of them is going to add the shockwave circles with this code.

Comment (EliphasNUIT) [reply GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs:933]: No scratch that comment, it is just the github diff hiding under which case that code was.

File: GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs [hunk 9]
```diff
@@ -963,6 +1222,378 @@ internal override void ComputeNPCCombatReplayActors(NPC target, ParsedEvtcLog lo
...
+                    break;
+                case (int)ArcDPSEnums.TrashID.VoidTimeCaster:
+                    // Gravity Crush - Indicator
+                    if (log.CombatData.TryGetEffectEventsByGUID(EffectGUIDs.HarvestTempleVoidTimecasterGravityCrushIndicator, out IReadOnlyList<EffectEvent> gravityCrushIndicators))
+                    {
+                        foreach (EffectEvent effect in gravityCrushIndicators)
+                        {
+                            int minRadius = 0;
+                            int radiusIncrease = 40;
+                            float initialOpacity = 1;
```
Comment (EliphasNUIT) [inline GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs:1259]: No opaque filled decorations, start at 0.7 or 0.8

File: GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs [hunk 10]
```diff
@@ -963,6 +1222,378 @@ internal override void ComputeNPCCombatReplayActors(NPC target, ParsedEvtcLog lo
...
+                            (long start, long end) lifespan = effect.ComputeLifespan(log, 1600);
+                            var positionConnector = new PositionConnector(effect.Position);
+                            // The indicator has 8 circles
+                            for (int i = 1; i <= 8; i++)
+                            {
+                                int maxRadius = minRadius + radiusIncrease;
+                                float opacity = initialOpacity / i;
+                                var circle = new DoughnutDecoration(minRadius, maxRadius, lifespan, Colors.Orange, opacity, positionConnector);
```
Comment (EliphasNUIT) [inline GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs:1267]: Prefer CircleDecoration when innerRadius is 0

File: GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs [hunk 11]
```diff
@@ -892,44 +992,197 @@ internal override void ComputeNPCCombatReplayActors(NPC target, ParsedEvtcLog lo
...
+                                    // Add projectile - Starts when the previous AoE ends because it's bouncing
+                                    (long start, long end) lifespanAnimation = (startingAoE.Time + startingAoE.Duration, endingAoE.Time + endingAoE.Duration);
+                                    replay.AddProjectile(startingAoE.Position, endingAoE.Position, lifespanAnimation, Colors.Black, 0.5);
+                                }
+                            }
                         }
                     }
+
+                    // Generic Orange AoE - Used in multiple sections of the encounter - Used for the void orbs during the tail and claw slams
+                    log.CombatData.TryGetEffectEventsByGUID(EffectGUIDs.HarvestTempleScalableOrangeAoE, out IReadOnlyList<EffectEvent> genericOrangeAoE);
```
Comment (EliphasNUIT) [inline GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs:1134]: Should be inside an if like the rest of TryGets, genericOrangeAoE list can be null.

File: GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs [hunk 12]
```diff
@@ -818,11 +863,33 @@ internal override void ComputeNPCCombatReplayActors(NPC target, ParsedEvtcLog lo
                             replay.Decorations.Add(new CircleDecoration(350, (start, end), Colors.Black, 0.4, new PositionConnector(effect.Position)));
                         }
                     }
+                    // Crystal Barrage
+                    if (log.CombatData.TryGetEffectEventsByGUID(EffectGUIDs.HarvestTempleKralkatorrikCrystalBarrageImpact, out IReadOnlyList<EffectEvent> crystalBarrage))
+                    {
+                        log.CombatData.TryGetEffectEventsByGUID(EffectGUIDs.HarvestTempleScalableOrangeAoE, out IReadOnlyList<EffectEvent> aoeIndicator);
```
Comment (EliphasNUIT) [inline GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/HarvestTemple.cs:869]: Should be also in an if

**Comment by EliphasNUIT** [comment]
Thanks!

**Pull request merged**